### ISSUE TITLE
Cache is disabled when output is updated

### ISSUE THREAD

**Issue by JohanWranker** [gradle/gradle#4316]
### Expected Behavior
Expect that the gradle cache is utilized to its full capability (i.e. used when a re-build is performed, updated when a new artifact is built)

### Current Behavior
If the task output is updated before the task is executed a full build is always executed and not pushed to the cache. 

### Context
To use the gradle-cache for long running operations (5-30 minutes) are real game-changers. The operation is code generation from models such as [Simulink](https://se.mathworks.com/help/dsp/ug/generate-c-code-from-simulink-model.html), [Autosar](https://vector.com/vi_autosar_tools_en.html) or [Fpga](https://en.wikipedia.org/wiki/Xilinx_Vivado).
* Those tools can (usually) be run in gui-mode and cli-mode; a normal workflow it to iterate thru update + generate + test in the gui and finalized by a cli-generation. The problem is that the final cli-generation is not stored in the gradle-cache since the iterations updated the output. 
* No tool is perfect and is it therefore common to patch some issue in the generated output, the patch would then update the original-output and thereby prevent a cache-operation in the next run.

In order to circumvent the cache-limitation we have setup a pre-task (to clean the output) and a port-task (to copy the output and perform the patches) but it less then perfect with the two generation modes gui vs cli and build-dependencies.

This can be solved by an annotation to the outputfile/outputdir that updated data is expected

### Steps to Reproduce
```bash
cat << EOF > build.gradle
@CacheableTask
class MyTask extends DefaultTask {
    @InputFile
  def inputFile = new File('inputFile')
  
  @OutputFile
  def outFile = new File('outFile')

  @TaskAction
  void execute(IncrementalTaskInputs inputs) {
    println 'Performing generation ... please hold'
    outFile.write('Generation data\n')
  }
}
    
task myTask(type: MyTask)

EOF

gradle wrapper --gradle-version 4.5

echo "input with 'random' data"
echo $$ > inputFile

echo "Run the generation once"
./gradlew --no-daemon --build-cache myTask

echo "sucessful use of cache"  
rm -f outFile
./gradlew --no-daemon --build-cache myTask
echo "****"

echo "FAIL1 when output is tainted and the cache is not used"
echo "tainted output" >> outFile
./gradlew --no-daemon --build-cache myTask
echo "****"

echo "Setup: The input new and update tainted"
echo "new input" >> inputFile
echo "tainted output" >> outFile
./gradlew --no-daemon --build-cache myTask

echo "FAIL2: Here gradle should be able to use cached info (but was not stored by the previous step)"
rm -f outFile
./gradlew --no-daemon --build-cache myTask
echo "****"
```
The output
```bash
input with 'random' data
Run the generation once
Build cache is an incubating feature.
:myTask
Performing generation ... please hold

BUILD SUCCESSFUL in 2s
1 actionable task: 1 executed
sucessful use of cache
Build cache is an incubating feature.
:myTask FROM-CACHE

BUILD SUCCESSFUL in 2s
1 actionable task: 1 from cache
****
FAIL1 when output is tainted and the cache is not used
Build cache is an incubating feature.
:myTask
Performing generation ... please hold

BUILD SUCCESSFUL in 2s
1 actionable task: 1 executed
****
Setup: The input new and update tainted
Build cache is an incubating feature.
:myTask
Performing generation ... please hold

BUILD SUCCESSFUL in 2s
1 actionable task: 1 executed
FAIL2: Here gradle should be able to use cached info (but was not stored by the previous step)
Build cache is an incubating feature.
:myTask
Performing generation ... please hold

BUILD SUCCESSFUL in 2s
1 actionable task: 1 executed
****

```

### Environment
Gradle 4.5 
Windows 8
Cygwin

**Comment by blindpirate** [comment]
Thanks for reporting, I'm not sure if this is a bug or not, but maybe @lptr can answer this.

**Comment by lptr** [comment]
Thanks @JohanWranker for reporting this, I managed to reproduce and add an integration test for it.

**Comment by lptr** [comment]
I looked into this a little more. The reason for the task not being cached is that we detect the tampering with the outputs. Unfortunately, at this point we can't decide if the tampering happened outside the build, or if it was caused by another task having produced overlapped outputs. If it were overlapping outputs, we'd be right to disable caching, because there's no good way to decide which task's outputs the overlapping outputs belong to. In your case we could simply overwrite the outputs, but we don't have enough information to detect this case. We have plans to record information about task outputs in a way that we will be able to make this distinction at some point, but this work is not scheduled yet.
